src/hotspot/share/runtime/handshake.cpp

```diff
@@ -562,3 +569,66 @@ HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* ma
                        pr_ret == HandshakeState::_succeeded ? "including" : "excluding", p2i(match_op));
   return pr_ret;
 }
+
+void HandshakeState::lock() {
+  _lock.lock_without_safepoint_check();
+}
+
+void HandshakeState::unlock() {
+  _lock.unlock();
+}
+
+// The careful dance between thread suspension and exit is handled here.
+// The caller has notified the agents that we are exiting, before we go
+// on, we must check for a pending external suspend request and honor it
+// in order to not surprise the thread that made the suspend request.
+void HandshakeState::thread_exit() {
+  assert(JavaThread::current() == _handshakee, "Must be");
+  while (true) {
+    {
+      MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+      if (!_handshakee->is_suspend_requested()) {
+        _handshakee->set_exiting();
+        // No more external suspends are allowed at this point.
+        return;
+      }
+    }
+    ThreadBlockInVM tbivm(_handshakee);
+    // We're done with this suspend request, but we have to loop around
+    // and check again. Eventually we will get the lock without a pending
+    // external suspend request and will be able to mark ourselves as
+    // exiting.
+  }
+}
+
+void HandshakeState::suspend_in_handshake() {
+  assert(Thread::current() == _handshakee, "should call from _handshakee");
+  assert(_lock.owned_by_self(), "Lock must be held");
+  assert(!_handshakee->has_last_Java_frame() || _handshakee->frame_anchor()->walkable(), "should have walkable stack");
+  JavaThreadState jts = _handshakee->thread_state();
+  while (_handshakee->_suspended) {
+    _handshakee->set_thread_state(_thread_blocked);
```

Comment:
Yes it can, but I did not want to go to blocked if we are not suspended.
You think we should go to block even if we are no longer suspened?

Reply:
In that case, you need to move the `_handshakee->set_thread_state(jts)`
into the while-loop so that you're matching up the set to blocked with
the restore to previous.

Reply:
I don't want the state to 'flicker', it should be stable blocked while we are suspended.
Therefore I set when we have left the while loop and know we are leaving the suspension.